Virtual destructors

tntmeijs opened the pull request:
## Description
Child classes have their destructors marked virtual now.

## Motivation and Context
This ensures destruction always starts at the bottom of the inheritance chain.

## How Has This Been Tested?
I tried to compile / run the code, and no issues occurred.

## Screenshots (if appropriate):
Not applicable.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.

tntmeijs committed:
Fixed destructors

[CHANGE] Destructors of inherited classes have been marked virtual.
[FIX] Include statement had to be renamed because the naming convention changed.

sainsay reviewed (changes requested):
change virtual back to override. virtual is not needed on a function inherited by the child class

tntmeijs committed:
Pull request changes

[CHANGE] Virtual has been removed from destructors.
[CHANGE] Child classes have been marked as final.

sainsay reviewed (approved):
👍
